### PR TITLE
GUVNOR-2173 - Asset Oracle: Change Impact prediction 

### PR DESCRIPTION
**Pull request by mrietveld**
Change Impact, Part I: building a reference infrastructure that we can query.

For an explanation of the changes, see https://github.com/droolsjbpm/kie-wb-common/pull/414

PR set:
- https://github.com/uberfire/uberfire/pull/455
- https://github.com/droolsjbpm/kie-wb-common/pull/414
- https://github.com/droolsjbpm/drools-wb/pull/257
- https://github.com/droolsjbpm/jbpm-designer/pull/375
- https://github.com/droolsjbpm/jbpm-form-modeler/pull/63

Also, this PR depends on https://github.com/droolsjbpm/jbpm/pull/511

### CODE REVIEW

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/BPMN2FileIndexer.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
+package org.jbpm.designer.server.indexing;
```
Comment (manstis) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/BPMN2FileIndexer.java]: License

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/BPMN2FileIndexer.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        return bpmn2Service.buildProcessDefinition(deploymentId, bpmnStr, null, false);
+    }
+
+    private List<Process> getProcesses(String bpmnStr, ClassLoader projectClassLoader) throws SAXException, IOException {
+        XmlProcessReader processReader = new XmlProcessReader( modules, projectClassLoader );
+        return processReader.read( new StringReader(bpmnStr) );
+    }
+
+    @Override
+    public KObjectKey toKObjectKey( Path path ) {
```
Comment (manstis) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/BPMN2FileIndexer.java]: I thought I read somewhere that this method should not be overridden and ideally it'd have been ```final```?

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/ProcessDefintionIndexVisitor.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
+package org.jbpm.designer.server.indexing;
```
Comment (manstis) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/ProcessDefintionIndexVisitor.java]: License

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/ProcessIndexVisitor.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
+package org.jbpm.designer.server.indexing;
```
Comment (manstis) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/ProcessIndexVisitor.java]: License

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java [hunk 5]
```diff
@@ -177,6 +182,108 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
...
+            String ruleFlowGroupName = (String) row.getValue();
+            if( ! ruleFlowGroupNames.contains(ruleFlowGroupName) ) {
+                // but only add the new ones
+                otherRuleFlowGroupNames.add( ruleFlowGroupName );
+            }
+        }
+        Collections.sort( otherRuleFlowGroupNames );
+
+        // divider between asset project/branch ruleflowgroup names and all ruleflowgroup names
+        ruleFlowGroupNames.add("********");
```
Comment (manstis) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java]: I can't see that the old way had any demarcation between the two sets of Rule Flow Group Names. If you're adding the distinction I'd assume something somewhere client-side would need tweaking to handle it too?

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java [hunk 6]
```diff
@@ -177,6 +182,108 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
...
+            String ruleFlowGroupName = (String) row.getValue();
+            if( ! dataTypeNames.contains(ruleFlowGroupName) ) {
+                // but only add the new ones
+                otherDataTypeNames.add( ruleFlowGroupName );
+            }
+        }
+        Collections.sort( otherDataTypeNames );
+
+        // divider between asset project/branch ruleflowgroup names and all ruleflowgroup names
+        dataTypeNames.add("********");
```
Comment (manstis) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java]: Same here.

**Comment by mrietveld** [comment]
@tsurdilo  Would you mind also reviewing this? Thanks! It will help to read droolsjbpm/kie-wb-common#414 first.

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java (hunk 5 above)
Comment (mrietveld) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java]: This was a suggestion by @tsurdilo -- Tiho, could you +1 this change?

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java (hunk 6 above)
Comment (mrietveld) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java]: @tsurdilo +1, please? :)

**Comment by mrietveld** [comment]
Fixed missing licenses.

**Comment by mrietveld** [comment]
jenkins retest this

**Comment by tsurdilo** [comment]
@mrietveld could i pls have couple days to test this locally to make sure there is no impact on ui? im in middle of bigger pr now so pls give me just couple days or if @jlindop has some time now he can test as well. thanks.

**Comment by mrietveld** [comment]
@tsurdilo no problem.

**Comment by mrietveld** [comment]
jenkins retest this

(*really* fixed the non-public method/errai marshalling problems.. )

**Comment by manstis** [comment]
@mrietveld This is still red.

**Comment by mrietveld** [comment]
Mr. Jenkins, the please of the retest would be green. Yes. ;)

**Comment by mrietveld** [comment]
There was once a Jenkins that ate tests. "Please, may I have some more?" it would ask and then retest.

**Comment by jlindop** [comment]
@mrietveld   This builds now, which is good, but I haven't been able to test it.
I've re-built all the repo's which still have unmerged commits for GUVNOR-2173 and built kie-wb-distributions/kie-wb, then deployed the wildfl10 war file to wildfly10. The war fails to start with this error:

16:17:16,835 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-4) MSC000001: Failed to start service jboss.deployment.unit."kie-wb-7.0.0-SNAPSHOT-wildfly10.war".WeldStartService: org.jboss.msc.service.StartException in service jboss.deployment.unit."kie-wb-7.0.0-SNAPSHOT-wildfly10.war".WeldStartService: Failed to start service
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1904)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type DefinitionService with qualifiers @Default
  at injection point [BackedAnnotatedField] @Inject protected org.jbpm.designer.server.indexing.BPMN2FileIndexer.bpmn2Service
  at org.jbpm.designer.server.indexing.BPMN2FileIndexer.bpmn2Service(BPMN2FileIndexer.java:0)

	at org.jboss.weld.bootstrap.Validator.validateInjectionPointForDeploymentProblems(Validator.java:359)
	at org.jboss.weld.bootstrap.Validator.validateInjectionPoint(Validator.java:281)
	at org.jboss.weld.bootstrap.Validator.validateGeneralBean(Validator.java:134)
	at org.jboss.weld.bootstrap.Validator.validateRIBean(Validator.java:155)
	at org.jboss.weld.bootstrap.Validator.validateBean(Validator.java:518)
	at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:68)
	at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:66)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:63)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:56)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at org.jboss.threads.JBossThread.run(JBossThread.java:320)

16:17:16,844 ERROR [org.jboss.as.controller.management-operation] (Controller Boot Thread) WFLYCTL0013: Operation ("deploy") failed - address: ([("deployment" => "kie-wb-7.0.0-SNAPSHOT-wildfly10.war")]) - failure description: {"WFLYCTL0080: Failed services" => {"jboss.deployment.unit.\"kie-wb-7.0.0-SNAPSHOT-wildfly10.war\".WeldStartService" => "org.jboss.msc.service.StartException in service jboss.deployment.unit.\"kie-wb-7.0.0-SNAPSHOT-wildfly10.war\".WeldStartService: Failed to start service
    Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type DefinitionService with qualifiers @Default
  at injection point [BackedAnnotatedField] @Inject protected org.jbpm.designer.server.indexing.BPMN2FileIndexer.bpmn2Service
  at org.jbpm.designer.server.indexing.BPMN2FileIndexer.bpmn2Service(BPMN2FileIndexer.java:0)
"}}

**Comment by mrietveld** [comment]
@jlindop Thanks, looking into it.

**Comment by mrietveld** [comment]
@jlindop Fixed. The problem was that the `jbpm-services-cdi` module, which produces the `DefinitionService` CDI impl, was not included in either `jbpm-designer-standalone` or the workbench.

**Comment by mrietveld** [comment]
Argh.. fixing. `jbpm-kie-services` includes `dashbuilder` jars, which mess up the build..

**Comment by mrietveld** [comment]
@jlindop Argh, needs another PR to fix this.

**Comment by csadilek** [comment]
@cristianonicolai as I will be off for a couple of days, can you merge this PR series once this one as a plus one and is green? Thanks.

**Comment by cristianonicolai** [comment]
@csadilek will do ;)

**Comment by jlindop** [comment]
@jomarko @tsurdilo  Will one of you test this when it's ready to test (Marco's working on it)? See above for the set of dependent PR's. I'm on PTO next week, and @mrietveld  is leaving on Aug31st, so we should try to merge it as soon as we can. Thanks, Jeremy.

**Comment by jomarko** [comment]
@jlindop I can't promise, but if I will have a moment I can try it. I will contact @tsurdilo when the PR will be ready and I will be out of time.

**Comment by mrietveld** [comment]
@jomarko @tsurdilo (and @mswiderski ?). This is done. If you have any questions, please let me know. 

( @manstis Once this is done, the rest can be merged as well? )

**Comment by mswiderski** [comment]
looks very good to me - from the process information capturing point of view. Thanks @mrietveld for implementing the listener in designer. I do believe that is the cleanest solution even though it does duplicate tiny bit of code.

**Comment by mrietveld** [comment]
@jomarko or @tsurdilo, would either of you have some time to review this? I have 3 other PR's waiting on this. 

Thanks!

**Comment by jomarko** [comment]
Here are my comments:

- `DesignerProcessDataEventListener` I did not find tests for this class and it seems to be as worth to be tested. Next there is typo error in variable `functionImpot`. It should be `functionImpot`. And the last question to this class. There is used is code `StringUtils.isEmpty()`. I looked at implementation of this and I think that it evaluates to false for cases like this `StringUtils.isEmpty(' ')`. Is it fine?

- `DesignerFindDataTypesQuery` and `FindRuleFlowNamesQuery`. In these classes I am wondering about methods `toQuery()` and `validateTerms()`. 

I think that in `DesignerFindDataTypesQuery.toQuery(terms)` is redundant check if `terms` are not null and empty.

Shouldn't be also in `DesignerFindDataTypesQuery.validateTerms(terms)` code that ensures that required term was found? Please compare `DesignerFindDataTypesQuery.validateTerms(terms)` with `FindRuleFlowNamesQuery.validateTerms(terms)` to see my point.

**Comment by mrietveld** [comment]
@jomarko Thanks!

- Tests: the test for the `DesignerProcessDataEventListener` class is this class: `BPMN2FileIndexerIndexingTest`. 
- `functionImpot`: will fix
- redundant check: I'll remove that
- `validateTerms`: will fix

Thanks.

**Comment by jomarko** [comment]
@mrietveld thanks, and the last thing. Is it fine if `StringUtils.isEmpty(' ')` evaluates to false in `DesignerProcessDataEventListener`?

**Comment by mrietveld** [comment]
@jomarko Yes, that's correct (it should evaluate to false and *not* to true!)

**Comment by mrietveld** [comment]
@jomarko Thanks for the feedback -- I've fixed the `validateTerms` and `toQuery` methods, and also added a test for the `validateTerms` method. 

(The `toQuery` is tested in the indexing test).

**Comment by mrietveld** [comment]
jenkins please retest

**Comment by jomarko** [comment]
+1

**Pull request merged**